tests/test_clean.py

```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
 
     # Test non-numeric entities
     ('&#', '&amp;#'),
-    ('&#<', '&amp;#&lt;')
+    ('&#<', '&amp;#&lt;'),
+
+    # html5lib tokenizer unescapes character entities, so these would become '
+    # and " which makes it possible to break out of html attributes.
+    #
+    # Verify that clean() doesn't unescape entities.
+    ('&#39;&#34;', '&#39;&#34;'),
 ])
 def test_character_entities(text, expected):
-    assert bleach.clean(text) == expected
+    assert clean(text) == expected
 
 
-def test_weird_strings():
-    s = '</3'
-    assert bleach.clean(s) == ''
-
-
-def test_stripping():
-    text = 'a test <em>with</em> <b>html</b> tags'
-    assert (
-        bleach.clean(text, strip=True) ==
+@pytest.mark.parametrize('data, kwargs, expected', [
+    # All tags are allowed, so it strips nothing
+    (
+        'a test <em>with</em> <b>html</b> tags',
+        {'strip': True},
         'a test <em>with</em> <b>html</b> tags'
-    )
+    ),
 
-    text = 'a test <em>with</em> <img src="http://example.com/"> <b>html</b> tags'
-    assert (
-        bleach.clean(text, strip=True) ==
+    # img tag is disallowed, so it's stripped
+    (
+        'a test <em>with</em> <img src="http://example.com/"> <b>html</b> tags',
+        {'strip': True},
         'a test <em>with</em>  <b>html</b> tags'
-    )
+    ),
 
-    text = '<p><a href="http://example.com/">link text</a></p>'
-    assert (
-        bleach.clean(text, tags=['p'], strip=True) ==
+    # a tag is disallowed, so it's stripped
+    (
+        '<p><a href="http://example.com/">link text</a></p>',
+        {'tags': ['p'], 'strip': True},
         '<p>link text</p>'
-    )
-    text = '<p><span>multiply <span>nested <span>text</span></span></span></p>'
-    assert (
-        bleach.clean(text, tags=['p'], strip=True) ==
+    ),
+
+    # handle nested disallowed tag
+    (
+        '<p><span>multiply <span>nested <span>text</span></span></span></p>',
+        {'tags': ['p'], 'strip': True},
         '<p>multiply nested text</p>'
-    )
+    ),
 
-    text = '<p><a href="http://example.com/"><img src="http://example.com/"></a></p>'
-    assert (
-        bleach.clean(text, tags=['p', 'a'], strip=True) ==
+    # handle disallowed tag that's deep in the tree
+    (
+        '<p><a href="http://example.com/"><img src="http://example.com/"></a></p>',
+        {'tags': ['a', 'p'], 'strip': True},
         '<p><a href="http://example.com/"></a></p>'
-    )
+    ),
+])
+def test_stripping_tags(data, kwargs, expected):
+    assert clean(data, **kwargs) == expected
+
+
+@pytest.mark.parametrize('data, expected', [
+    (
+        '<scri<script>pt>alert(1)</scr</script>ipt>',
+        'pt&gt;alert(1)ipt&gt;'
+    ),
+    (
+        '<scri<scri<script>pt>pt>alert(1)</script>',
+        'pt&gt;pt&gt;alert(1)'
+    ),
```

Comment:
As a side note, the output here is safe, but I don't know exactly how it goes from the `in` to the `out`. This is definitely curious and maybe worth looking into at some point.